### PR TITLE
[Wait for #1087] [Ini] Propose ini wrapper 2 @open sesame 04/22 13:18

### PR DESCRIPTION
**Pull request by zhoonit**
- [Ini] Propose ini wrapper 2 

```
Propose ini wrapper in nntrainer, most codes are from
`nntrainer_test_util.cpp`.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: nntrainer/utils/ini_wrapper.cpp [hunk 1]
```diff
@@ -68,8 +68,56 @@ void IniSection::setEntry(const std::string &entry_str) {
     int status = getKeyValue(cur, key, value);
     NNTR_THROW_IF(status != ML_ERROR_NONE, std::invalid_argument)
       << "getKeyValue Failed";
+
```
Comment (zhoonit) [inline nntrainer/utils/ini_wrapper.cpp:71]: This file is mostly moving code, so labeled `trivial`

File: nntrainer/utils/ini_wrapper.cpp [hunk 2]
```diff
@@ -69,8 +69,56 @@ void IniSection::setEntry(const std::string &entry_str) {
...
+  auto section = std::find_if(sections.begin(), sections.end(),
+                              [&](const IniSection &section) {
+                                return section.getName() == s.getName();
+                              });
+
+  NNTR_THROW_IF(section == sections.end(), std::invalid_argument)
+    << "section key is not found key: " << s.getName();
+
+  (*section) += s;
```
Comment (jijoongmoon) [inline nntrainer/utils/ini_wrapper.cpp:103]: is it append or overwrite?

File: nntrainer/utils/ini_wrapper.h [hunk 3]
```diff
@@ -173,6 +176,162 @@ class IniSection {
...
+  }
+
+private:
+  /**
+   * @brief update a section from a formatted string, `sectionkey / propkey=val
+   * | propkey=val`
+   *
+   * @param string_representation "model/optimizer=SGD | ..."
+   */
+  void updateSection(const std::string &string_representation);
```
Comment (kparichay) [inline nntrainer/utils/ini_wrapper.h:314]: how about std::tuple<std::string, std::string> than a joined version with `/`?

Comment (kparichay) [reply nntrainer/utils/ini_wrapper.h:314]: tuple/pair/array - whichever you like.

Comment (zhoonit) [reply nntrainer/utils/ini_wrapper.h:315]: I like to update, added as a todo for now as this won't be used for the serializer.

Comment (kparichay) [reply nntrainer/utils/ini_wrapper.h:315]: ok.

**Review by kparichay — APPROVED** [review]
LGTM

File: nntrainer/utils/ini_wrapper.cpp (hunk 2 above)
Comment (zhoonit) [reply nntrainer/utils/ini_wrapper.cpp:103]: If key clashes, this is overwrite, if key doesn't clash this is append

File: nntrainer/utils/ini_wrapper.h [hunk 4]
```diff
@@ -176,6 +179,162 @@ class IniSection {
...
+
+  /**
+   * @brief save ini to a file
+   */
+  void save_ini();
+  /**
+   * @brief erase ini
+   *
+   */
+  void erase_ini() noexcept { remove(getIniName().c_str()); }
```
Comment (lhs8928) [inline nntrainer/utils/ini_wrapper.h:297]: Please check a return value

**Review by lhs8928 — APPROVED** [review]
LGTM

File: nntrainer/utils/ini_wrapper.h (hunk 4 above)
Comment (zhoonit) [reply nntrainer/utils/ini_wrapper.h]: Done

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**